### PR TITLE
Redmine query parameters

### PR DESCRIPTION
**Pull request by CaptainQuirk**
:dart: Goals
---

- Provide a mechanism to filter issues

:panda_face: Implementation
---

- Support a query configuration value to filter issues

:notebook_with_decorative_cover: Documentation
---

- document the use of the query setting and how it overrides other configuration value
- link to redmine issues api documentation

### CODE REVIEW

File: bugwarrior/services/redmine.py [hunk 1]
```diff
@@ -31,16 +32,21 @@ def __init__(self, url, key, auth, issue_limit, verify_ssl):
...
 
-    def find_issues(self, issue_limit, only_if_assigned=False):
+    def find_issues(self, issue_limit, query, only_if_assigned=False):
         args = {}
         # TODO: if issue_limit is greater than 100, implement pagination to return all issues.
         # Leave the implementation of this to the unlucky soul with >100 issues assigned to them.
-        if issue_limit is not None:
-            args["limit"] = issue_limit
+        url = "/issues.json"
+        if query is not None:
```
Comment (ryneeverett) [inline bugwarrior/services/redmine.py]: Isn't the default value an empty string, rather than None?

File: bugwarrior/services/redmine.py [hunk 2]
```diff
@@ -31,16 +32,21 @@ def __init__(self, url, key, auth, issue_limit, verify_ssl):
...
+            url = url + "?" + query
+        else:
+            if issue_limit is not None:
```
Comment (ryneeverett) [inline bugwarrior/services/redmine.py]: The comments regarding `issue_limit` should remain above this line.

File: bugwarrior/docs/services/redmine.rst [hunk 3]
```diff
@@ -28,6 +28,22 @@ If you want to ignore verifying the SSL certificate, set::
...
+
+If you need to take control over the filtering of issues, you may use the
+`redmine.query` setting.
+
+Be aware though that the `redmine.issue_limit` and `redmine.only_if_assigned`
+settings will be then be ignored.
+
+You'll have to include their redmine equivalent in the query::
+
+    redmine.query = project_id=10&status_id=15&limit=100&assigned_to_id=me
```
Comment (ryneeverett) [inline bugwarrior/docs/services/redmine.rst]: Sorry if I misled you regarding the extent to which `query` should override default behavior -- I hadn't realized how much simpler the Redmine service is than the others in which we've implemented this option. In this case I don't think it should override either of these configuration options and therefore you could just add it to the conventional querystring as I believe you first suggested.

File: bugwarrior/services/redmine.py (hunk 1 above)
Comment (CaptainQuirk) [reply bugwarrior/services/redmine.py]: You're right ! Solved it

File: bugwarrior/services/redmine.py (hunk 2 above)
Comment (CaptainQuirk) [reply bugwarrior/services/redmine.py]: Done !

File: bugwarrior/docs/services/redmine.rst (hunk 3 above)
Comment (CaptainQuirk) [reply bugwarrior/docs/services/redmine.rst]: I handled it the way I first imagined it.

**Comment by CaptainQuirk** [comment]
I also saw that a suite of github actions failed on my branch : https://github.com/ralphbean/bugwarrior/actions/runs/2031120214. I used pytest locally and got a 0 exit code (and a bunch of warnings about distutils version classes being deprecated, due to the fact that python is installed with homebrew on my Linux Mint)

**Comment by ryneeverett** [comment]
It's not clear to me that there is any need for an `if`/`else` clause here. Couldn't we continue passing the limit and assigned_to_id parameters in `args` even with additional query parameters in the url?

**Comment by CaptainQuirk** [comment]
How should I handle the `query` setting which is a partial query string, then ?
Due to its partial nature, I'm not sure I could parse it into a dict easily.

**Comment by CaptainQuirk** [comment]
Maybe your thought was to both append the string value of the query setting to the url variable while also filling the arg dict ?

**Comment by ryneeverett** [comment]
> Maybe your thought was to both append the string value of the query setting to the url variable while also filling the arg dict ?

Yes, that was my thought and I think it's the best route unless there is some reason it doesn't work.

> How should I handle the query setting which is a partial query string, then ?
Due to its partial nature, I'm not sure I could parse it into a dict easily.

I actually think it is fairly easy to parse a query string into a dict with [urlparse.parse_qs](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.parse_qs).

However, if for some reason we're forced to choose between strings and dicts, I think it would be more reliable to serialize the limit and and assigned_to_id like you did than to parse the user's query string which is just another opportunity for error. However, I'd want to serialize it every time, not just when `query != ''`, so we still wouldn't need the `else` clause. We want to minimize logical branching. Make sense?

**Comment by CaptainQuirk** [comment]
> I actually think it is fairly easy to parse a query string into a dict with urlparse.parse_qs.

A full blown query string, maybe, but something like `project_id=10&status_id=15` with no leading `&` or `?`, I'm not sure.

So to wrap it up  : 

- Try to add the append the given query setting to the url, keeping the `assigned_to_id` and `limit` parameters in the arg dict
- If it does not work, serialize `limit`, serialize `assigned_to_id` if `only_if_assigned` is set and append the query, if it is not the empty string

That should indeed reduce the branching !

**Comment by ryneeverett** [comment]
I think we could completely remove the branching by appending `query` even if it is an empty string because I believe a trailing ampersand is acceptable.

**Comment by CaptainQuirk** [comment]
That should do it now !

**Review by ryneeverett — APPROVED** [review]

**Pull request merged**